cli/src/semgrep/rule_match.py

```diff
@@ -304,19 +304,45 @@ def is_blocking(self) -> bool:
 
     @property
     def dataflow_trace(self) -> Optional[core.CliMatchDataflowTrace]:
+        def translate_loc(location: core.Location) -> Tuple[core.Location, str]:
+            with open(location.path, errors="replace") as fd:
+                content = util.read_range(
+                    fd, location.start.offset, location.end.offset
+                )
+            return (location, content)
+
+        def translate_core_match_call_trace(
+            call_trace: core.CoreMatchCallTrace,
+        ) -> core.CliMatchCallTrace:
+            if isinstance(call_trace.value, core.CoreLoc):
+                return core.CliMatchCallTrace(
+                    core.CliLoc(translate_loc(call_trace.value.value))
+                )
+            elif isinstance(call_trace.value, core.CoreCall):
+                intermediate_vars = []
+                for var in call_trace.value.value[1]:
+                    location, content = translate_loc(var.location)
+                    intermediate_vars.append(
+                        core.CliMatchIntermediateVar(location=location, content=content)
+                    )
```

Comment:
This looks to me like it could be a list comprehension, I think we generally prefer those where possible for purity + speed reasons